apps/ejabberd/src/mongoose_cassandra_worker.erl

```diff
@@ -0,0 +1,494 @@
+%%%-------------------------------------------------------------------
+%%% @author Uvarov Michael <arcusfelis@gmail.com>
+%%% @copyright 2016 Erlang Solutions, Ltd.
+%%% @doc Generic cassandra worker
+%%% @end
+%%%-------------------------------------------------------------------
+-module(mongoose_cassandra_worker).
+
+%% ----------------------------------------------------------------------
+%% Exports
+
+%% API
+-export([cql_query/5,
+         cql_query_pool/5,
+         cql_query_async/5,
+         cql_query_pool_async/5,
+         cql_query_multi_async/5,
+         cql_query_pool_multi_async/5,
+         cql_batch/4,
+         cql_batch_pool/4,
+         cql_batch/5,
+         cql_batch_pool/5]).
+
+%% Helpers for debugging
+-export([test_query/1,
+         test_query/2,
+         queue_length/1,
+         queue_lengths/1,
+         total_count_query/2]).
+
+%% Internal exports
+-export([start_link/4]).
+
+%% mongoose_cassandra_worker callbacks
+-export([prepared_queries/0]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-behaviour(gen_server).
+-behaviour(mongoose_cassandra).
+
+%% ----------------------------------------------------------------------
+%% Imports
+
+-include_lib("ejabberd/include/ejabberd.hrl").
+-include_lib("ejabberd/include/jlib.hrl").
+-include_lib("exml/include/exml.hrl").
+
+-record(state, {
+    pool_name,
+    conn,
+    prepared_queries,
+    query_refs,
+    query_refs_count}).
+
+-record(prepared_query, {
+          query_id,
+          query_types}).
+
+%%====================================================================
+%% Internal functions
+%%====================================================================
+
+start_link(PoolName, Addr, Port, ClientOptions) ->
+    gen_server:start_link(?MODULE, [PoolName, Addr, Port, ClientOptions], []).
+
+%% @doc Do CQL query
+%%
+%% UserJID is used for rate-limiting, statistics and debugging
+cql_query(Worker, _UserJID, Module, QueryName, Params) when is_pid(Worker) ->
+    try gen_server:call(Worker, {cql_query, Module, QueryName, Params}) of
+        {error, Reason} ->
+            {error, Reason};
+        ResultF ->
+            {ok, Result} = ResultF(),
+            case Result of
+                void ->
+                    {ok, []};
+                _ ->
+                    {ok, seestar_result:rows(Result)}
+            end
+    catch Class:Reason ->
+          {error, {Class, Reason}}
+    end.
+
+cql_query_async(Worker, _UserJID, Module, QueryName, Params) when is_pid(Worker) ->
+    gen_server:cast(Worker, {async_cql_query, Module, QueryName, Params}).
+
+cql_query_multi_async(Worker, _UserJID, Module, QueryName, MultiParams)
+  when is_pid(Worker) ->
+    gen_server:cast(Worker, {multi_async_cql_query, Module, QueryName, MultiParams}).
+
+cql_batch(Worker, UserJID, Module, Queries, not_batch) ->
```

Comment:
Batch is a feature of the driver to run queries on the same partition.
There are several types of them, logged (it's like a transaction) and not logged.

Batch can access data from different partitions but than cassandra documentation recommends to execute each query on the client side for performance reasons. 
So, not_batch is exactly what it is "execute these queries not using server-side batch but keep the same API".

So, we can name it `maybe_logged_or_not_or_manual_batch` but it feels like java.